### PR TITLE
Fs3 mc deals

### PR DESCRIPTION
**Pull request by GuohaoMa**
Add online deal feature to FS3-mc and update readme.

### CODE REVIEW

**Comment by flyworker** [comment]
@GuohaoMa it has conflict please resolve it first

File: cmd/import-main.go [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+		Name:   "object",
+		Usage:  "specify which object to import",
+	},
+}
+
+func checkImportArgs(ctx *cli.Context) (string, string) {
+	args := ctx.Args()
+	for _, arg := range args {
+		if strings.TrimSpace(arg) == "" {
+			fatalIf(errInvalidArgument().Trace(args...), "Unable to validate empty argument.")
```
Comment (zzhao-nbai) [inline cmd/import-main.go:58]: add return  here

File: cmd/import-main.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+		}
+	}
+	if len(args) < 1 {
+
```
Comment (zzhao-nbai) [inline cmd/import-main.go]: should add sth or remove

File: cmd/import-main.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+	}
+
+	bucket := strings.TrimSpace(ctx.String("bucket"))
+	object := strings.TrimSpace(ctx.String("object"))
+	if len(bucket) == 0 {
+		fatalIf(errInvalidArgument(), "please provide a valid bucket name")
```
Comment (zzhao-nbai) [inline cmd/import-main.go:65]: add retrun

File: cmd/import-main.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+	}
+	if len(object) == 0 {
+		fatalIf(errInvalidArgument(), "please provide a valid objective name")
```
Comment (zzhao-nbai) [inline cmd/import-main.go:69]: add return

File: cmd/import-main.go [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+	dataCID, err := ExecCommand(commandLine)
+	if err != nil {
+		logs.GetLogger().Error(err)
+		return
+	}
+	outStr := strings.Fields(string(dataCID))
+	dataCIDStr := outStr[len(outStr)-1]
+
+	if err != nil {
+		errorIf(errDummy(), err.Error())
```
Comment (zzhao-nbai) [inline cmd/import-main.go]: use log mentod and return

File: cmd/import-main.go [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+	if len(object) == 0 {
+		fatalIf(errInvalidArgument(), "please provide a valid objective name")
+	}
+
+	return bucket,object
+}
+
+func proposeImportData(config ImportOnlineDealData) {
+
+	//var commandArgs []string
```
Comment (zzhao-nbai) [inline cmd/import-main.go]: remove comment that do not need

File: cmd/import-main.go [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+	}
+	outStr := strings.Fields(string(dataCID))
+	dataCIDStr := outStr[len(outStr)-1]
+
+	if err != nil {
+		errorIf(errDummy(), err.Error())
+	} else {
+		fmt.Println(fmt.Sprintf("Bucket: %s, Object: %s, DataCid: %s", config.Bucket, config.Object, dataCIDStr))
+	}
+}
```
Comment (zzhao-nbai) [inline cmd/import-main.go:91]: add return at last

File: cmd/send-main.go [hunk 8]
```diff
@@ -215,6 +298,38 @@ func checkSendArgs(ctx *cli.Context) (string, uint, uint, string, string, *big.F
 	return wallet, start, duration, miner, input, priceDecimal
 }
 
+func checkSendOnlineArgs(ctx *cli.Context) (string, string,string, string, string,string, string) {
+	args := ctx.Args()
+	for _, arg := range args {
+		if strings.TrimSpace(arg) == "" {
+			fatalIf(errInvalidArgument().Trace(args...), "Unable to validate empty argument.")
```
Comment (zzhao-nbai) [inline cmd/send-main.go:305]: add return

File: cmd/send-main.go [hunk 9]
```diff
@@ -215,6 +298,38 @@ func checkSendArgs(ctx *cli.Context) (string, uint, uint, string, string, *big.F
...
+		}
+	}
+	if len(args) < 1 {
+
```
Comment (zzhao-nbai) [inline cmd/send-main.go]: add sth or remove

File: cmd/send-main.go [hunk 10]
```diff
@@ -215,6 +298,38 @@ func checkSendArgs(ctx *cli.Context) (string, uint, uint, string, string, *big.F
...
+	wallet := strings.TrimSpace(ctx.String("from"))
+	verifiedDeal := strings.TrimSpace(ctx.String("verified-deal"))
+	fastRetrieval := strings.TrimSpace(ctx.String("fast-retrieval"))
+	datacid := ctx.String("data-cid")
+	miner := ctx.String("miner-id")
+	duration := ctx.String("duration")
+	price := ctx.String("price")
+
+	if len(wallet) == 0 {
+		fatalIf(errInvalidArgument().Trace(wallet), "please provide a valid wallet")
```
Comment (zzhao-nbai) [inline cmd/send-main.go]: add return

File: cmd/send-main.go [hunk 11]
```diff
@@ -215,6 +298,38 @@ func checkSendArgs(ctx *cli.Context) (string, uint, uint, string, string, *big.F
...
+	}
+	if len(duration) == 0 {
+		fatalIf(errInvalidArgument(), "please provide a valid length of duration in day")
```
Comment (zzhao-nbai) [inline cmd/send-main.go:326]: add return

File: cmd/send-main.go [hunk 12]
```diff
@@ -215,6 +298,38 @@ func checkSendArgs(ctx *cli.Context) (string, uint, uint, string, string, *big.F
...
+	}
+	if len(price) ==0 {
+		fatalIf(errInvalidArgument(), "please provide a valid price")
+	}
```
Comment (zzhao-nbai) [inline cmd/send-main.go:332]: add return

File: cmd/send-main.go [hunk 13]
```diff
@@ -259,6 +374,27 @@ func proposeOfflineDeal(config *OfflineDeal) {
 	}
 }
 
+func proposeOnlineDeal(config OnlineDeal) {
+
+	//var commandArgs []string
+	//commandArgs = []string{"client", "deal", "--from", config.SenderWallet, "--verified-deal=", config.VerifiedDeal,
```
Comment (zzhao-nbai) [inline cmd/send-main.go]: remove comment you do not need any more

File: cmd/send-main.go [hunk 14]
```diff
@@ -259,6 +374,27 @@ func proposeOfflineDeal(config *OfflineDeal) {
...
+	//dealCID, err := exec.Command("lotus",commandArgs...).Output()
+
+	commandLine := "lotus " + "deal " + "--from " + config.SenderWallet + " --verified-deal="+config.VerifiedDeal+
+		" --fast-retrieval="+config.FastRetrieval+" "+config.DataCid+" "+config.MinerId+" "+config.Cost+" "+config.Duration
+	dealCID, err := ExecCommand(commandLine)
+	if err != nil {
+		logs.GetLogger().Error(err)
+		return
+	} else {
+		fmt.Println(fmt.Sprintf("DataCid: %s, DealCid: %s", config.DataCid, strings.TrimSuffix(string(dealCID),"\n")))
```
Comment (zzhao-nbai) [inline cmd/send-main.go]: add return

File: cmd/send-main.go [hunk 15]
```diff
@@ -380,3 +516,22 @@ func uploadCsv(csvPath string, targetFolder string, cliCtx *cli.Context) {
 		fatalIf(probe.NewError(e).Untrace(), e.Error())
```
Comment (zzhao-nbai) [inline cmd/send-main.go:513]: add return

File: cmd/send-main.go [hunk 16]
```diff
@@ -380,3 +516,22 @@ func uploadCsv(csvPath string, targetFolder string, cliCtx *cli.Context) {
...
+		return "", err
+	}
+	out_bytes, _ := ioutil.ReadAll(stdout)
+	if err := stdout.Close(); err != nil {
+		logs.GetLogger().Error("Execute failed when close stdout:" + err.Error())
+		return "", err
+	}
+	if err := cmd.Wait(); err != nil {
+		logs.GetLogger().Error("Execute failed when Wait:" + err.Error())
+		//return "", err
```
Comment (zzhao-nbai) [inline cmd/send-main.go]: remove //

File: logs/logs_logger.go [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+	Log.SetLevel(log.InfoLevel)
+	formatter := &log.TextFormatter{
+		TimestampFormat: "2006-01-02 15:04:05.000",
+		FullTimestamp:   true,
+	}
+	Log.SetReportCaller(true)
+	Log.SetFormatter(formatter)
+	pathMap := lfshook.PathMap{
+		log.InfoLevel:  "./logs/ledger_info.log",
+		log.WarnLevel:  "./logs/ledger_warn.log",
```
Comment (zzhao-nbai) [inline logs/logs_logger.go]: change ledger to fs3_mc

File: cmd/import-main.go (hunk 1 above)
Comment (GuohaoMa) [reply cmd/import-main.go:58]: Done

File: cmd/import-main.go (hunk 2 above)
Comment (GuohaoMa) [reply cmd/import-main.go]: Done

File: cmd/import-main.go (hunk 3 above)
Comment (GuohaoMa) [reply cmd/import-main.go:65]: Done.

File: cmd/import-main.go (hunk 4 above)
Comment (GuohaoMa) [reply cmd/import-main.go:69]: Done.

File: cmd/import-main.go (hunk 5 above)
Comment (GuohaoMa) [reply cmd/import-main.go]: Done.

File: cmd/import-main.go (hunk 6 above)
Comment (GuohaoMa) [reply cmd/import-main.go]: Done.

File: cmd/import-main.go (hunk 7 above)
Comment (GuohaoMa) [reply cmd/import-main.go:91]: Done.

File: cmd/send-main.go (hunk 8 above)
Comment (GuohaoMa) [reply cmd/send-main.go:305]: Done.

File: cmd/send-main.go (hunk 9 above)
Comment (GuohaoMa) [reply cmd/send-main.go]: Done.

File: cmd/send-main.go (hunk 10 above)
Comment (GuohaoMa) [reply cmd/send-main.go]: Done.

File: cmd/send-main.go (hunk 11 above)
Comment (GuohaoMa) [reply cmd/send-main.go:326]: Done.

File: cmd/send-main.go (hunk 12 above)
Comment (GuohaoMa) [reply cmd/send-main.go:332]: Done.

File: cmd/send-main.go (hunk 13 above)
Comment (GuohaoMa) [reply cmd/send-main.go]: Done.

File: cmd/send-main.go (hunk 14 above)
Comment (GuohaoMa) [reply cmd/send-main.go]: Done.

File: cmd/send-main.go (hunk 15 above)
Comment (GuohaoMa) [reply cmd/send-main.go:513]: Done.

File: cmd/send-main.go (hunk 16 above)
Comment (GuohaoMa) [reply cmd/send-main.go]: Done.

File: logs/logs_logger.go (hunk 17 above)
Comment (GuohaoMa) [reply logs/logs_logger.go]: Done.

**Pull request merged**